File: impl/kotlin/src/test/kotlin/AudienceKatTest.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.io.File

/**
 * The object-audience (field 13, §2.5.3) known-answer + gate test for the Kotlin SDK, in the same
 * plain-main() KAT style as WorkedExampleKatTest (no test framework).
 *
 * Three properties, all mutation-surviving:
 *  1. BYTE MATCH -- an audience-bearing object reproduces the independent oracle's content id AND
 *     payload (over the body WITH the audience), and a NO-audience object reproduces the base
 *     content id -- proving omit-when-empty additivity byte-for-byte
 *     (Kotlin == Go == Rust == ... == the oracle). contentId()/bodyMap() run the impl encoder.
 *  2. checkAudience -- the three-branch point-of-use gate.
 *  3. consumeObject -- enforced at the consume choke point BEFORE the compare-and-set: wrong/absent
 *     -> WrongAudience with no append; unnamed ledger -> LedgerUnsigned; correct -> consumes once.
 *
 * Run: compile the SDK sources plus this file, then
 *   java -cp "aud-kat.jar;<bcprov.jar>;<kotlin-stdlib.jar>" sh.bubblefish.naalp.AudienceKatTestKt
 */

private val A_SIGNER = Hex.decode("5349474e45525f41") // "SIGNER_A"
private const val A_AUDIENCE = "consuming-authority-xyz"

private class AudFailure(msg: String) : RuntimeException(msg)

private fun aeq(what: String, got: Any?, want: Any?) {
    if (got != want) throw AudFailure("$what:\n  got  = $got\n  want = $want")
}

private fun audienceObject(): Envelope.Object =
    Envelope.Object(
        kind = 2, channel = 4, tier = 0, signer = A_SIGNER, created = 1785000000000L,
        effect = 2, profile = 1, body = Cbor.T("hello"), audience = A_AUDIENCE,
    )

private fun plainObject(): Envelope.Object =
    Envelope.Object(
        kind = 2, channel = 4, tier = 0, signer = A_SIGNER, created = 1785000000000L,
        effect = 2, profile = 1, body = Cbor.T("hello"),
    )

private fun gateObject(aud: String): Envelope.Object =
    Envelope.Object(
        kind = 2, channel = 4, signer = A_SIGNER, created = 0, effect = 0,
        body = Cbor.T("x"), profile = 1, audience = aud,
    )

private fun findEnvelopeVector(): File? {
    var d: File? = File(".").absoluteFile
    repeat(8) {
        val cur = d ?: return null
        val p = File(cur, "vectors/envelope/cases.json")
        if (p.isFile) return p
        d = cur.parentFile
    }
    return null
}

/** hex value of [key] in the object_with_audience block (audience=true) or the base object block. */
private fun hexInBlock(json: String, audience: Boolean, key: String): String {
    val split = json.indexOf("object_with_audience")
    val region = if (audience) json.substring(split) else json.substring(0, split)
    val m = Regex("\"$key\"\\s*:\\s*\"([0-9a-f]+)\"").find(region)
        ?: throw AudFailure("$key not found in ${if (audience) "audience" else "base"} block")
    return m.groupValues[1]
}

private fun expectKind(what: String, wantKind: String, fn: () -> Unit) {
    try {
        fn()
        throw AudFailure("$what: expected $wantKind, no exception thrown")
    } catch (e: NaalpException) {
        aeq("$what.kind", e.kind, wantKind)
    }
}

private fun aid(): ByteArray = ByteArray(50) { it.toByte() }

fun main() {
    println("AudienceKatTest — object audience field 13 (§2.5.3): byte-match + gates")

    // 1. BYTE MATCH vs the independent oracle
    val vec = findEnvelopeVector()
    if (vec != null) {
        val json = vec.readText(Charsets.UTF_8)
        val a = audienceObject()
        a.id = a.contentId()
        aeq("audience content_id", Hex.encode(a.contentId()), hexInBlock(json, true, "content_id_hex"))
        aeq("audience payload", Hex.encode(Cbor.encode(a.bodyMap(true))), hexInBlock(json, true, "payload_hex"))
        aeq("plain content_id (additive)", Hex.encode(plainObject().contentId()), hexInBlock(json, false, "content_id_hex"))
        if (Hex.encode(audienceObject().contentId()) == Hex.encode(plainObject().contentId())) {
            throw AudFailure("audience and no-audience content ids must differ")
        }
        println("  ok  byte-match vs oracle (audience content_id + payload + additive plain content_id)")
    } else {
        println("  ..  vectors/envelope/cases.json not found; skipped the byte compare")
    }

    // 2. checkAudience three branches (6 cases)
    Envelope.checkAudience(gateObject("authority-A"), "authority-A", true)   // pass
    Envelope.checkAudience(gateObject(""), "authority-A", false)             // pass
    Envelope.checkAudience(gateObject("authority-A"), "authority-A", false)  // pass
    expectKind("consume-once foreign", "WrongAudience") { Envelope.checkAudience(gateObject("authority-B"), "authority-A", true) }
    expectKind("consume-once absent", "WrongAudience") { Envelope.checkAudience(gateObject(""), "authority-A", true) }
    expectKind("foreign unrestricted", "WrongAudience") { Envelope.checkAudience(gateObject("authority-B"), "authority-A", false) }
    println("  ok  checkAudience three branches (6 cases)")

    // 3. consumeObject choke point (a fresh WAL per case)
    run {
        val p = File.createTempFile("naalp-audience", ".wal")
        val led = Approval.openLedger(p.absolutePath, "authority-A")
        expectKind("consumeObject wrong", "WrongAudience") { led.consumeObject(gateObject("authority-B"), aid(), "consumer") }
        aeq("wrong audience no append", led.size(), 0)
        led.close(); p.delete()
    }
    run {
        val p = File.createTempFile("naalp-audience", ".wal")
        val led = Approval.openLedger(p.absolutePath, "authority-A")
        expectKind("consumeObject absent", "WrongAudience") { led.consumeObject(gateObject(""), aid(), "consumer") }
        aeq("absent audience no append", led.size(), 0)
        led.close(); p.delete()
    }
    run {
        val p = File.createTempFile("naalp-audience", ".wal")
        val led = Approval.openLedger(p.absolutePath, "authority-A")
        led.consumeObject(gateObject("authority-A"), aid(), "consumer")
        aeq("correct audience consumes once", led.size(), 1)
        expectKind("consumeObject second", "AlreadyConsumed") { led.consumeObject(gateObject("authority-A"), aid(), "consumer") }
        led.close(); p.delete()
    }
    run {
        val p = File.createTempFile("naalp-audience", ".wal")
        val led = Approval.openLedger(p.absolutePath, "")
        expectKind("consumeObject unnamed", "LedgerUnsigned") { led.consumeObject(gateObject("authority-A"), aid(), "consumer") }
        led.close(); p.delete()
    }
    println("  ok  consumeObject choke point (wrong/absent/correct/unnamed)")

    println("AudienceKatTest: PASS")
}

File: impl/csharp/test/AudienceKatTest.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// The object-audience (field 13, §2.5.3) known-answer + gate tests for the C# SDK.
    ///
    /// <para>Byte match: an audience-bearing object's content id (computed over the body WITH the
    /// audience) reproduces the independent oracle's <c>object_with_audience.content_id_hex</c>
    /// (vectors/envelope/cases.json), and a NO-audience object reproduces the base
    /// <c>object.content_id_hex</c> — proving omit-when-empty additivity byte-for-byte
    /// (C# == Go == Rust == Python == TS == Ruby == PHP == the oracle). ContentId() runs the impl's
    /// own BodyMap-with-audience, so a wrong field number/type/placement would flip it.</para>
    ///
    /// <para>CheckAudience is the three-branch point-of-use gate; ConsumeObject enforces it at the
    /// consume choke point BEFORE the compare-and-set (wrong/absent -> WrongAudience with no append;
    /// unnamed ledger -> LedgerUnsigned; correct -> consumes once).</para>
    /// </summary>
    public sealed class AudienceKatTest
    {
        private const int Alg = Cose.ALG_MLDSA65;
        private static readonly byte[] Signer = Convert.FromHexString("5349474e45525f41"); // "SIGNER_A"
        private const string Audience = "consuming-authority-xyz";

        private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();

        private static Envelope.Object AudienceObject() => new Envelope.Object(
            kind: 2, channel: 4, signer: Signer, created: 1785000000000L, effect: 2,
            body: new Cbor.T("hello"), tier: 0, profile: 1, audience: Audience);

        private static Envelope.Object PlainObject() => new Envelope.Object(
            kind: 2, channel: 4, signer: Signer, created: 1785000000000L, effect: 2,
            body: new Cbor.T("hello"), tier: 0, profile: 1);

        private static Envelope.Object GateObject(string aud) => new Envelope.Object(
            kind: 2, channel: 4, signer: Signer, created: 0, effect: 0,
            body: new Cbor.T("x"), profile: 1, audience: aud);

        private static string? FindVector()
        {
            var d = new DirectoryInfo(AppContext.BaseDirectory);
            for (int i = 0; i < 10 && d != null; i++)
            {
                string p = Path.Combine(d.FullName, "vectors", "envelope", "cases.json");
                if (File.Exists(p))
                {
                    return p;
                }
                d = d.Parent;
            }
            return null;
        }

        [Fact]
        public void AudienceContentIdMatchesOracle()
        {
            string? p = FindVector();
            Assert.True(p != null, "committed oracle vector vectors/envelope/cases.json not found from " + AppContext.BaseDirectory);
            using JsonDocument doc = JsonDocument.Parse(File.ReadAllText(p!, Encoding.UTF8));
            JsonElement root = doc.RootElement;
            string wantAudience = root.GetProperty("object_with_audience").GetProperty("content_id_hex").GetString()!;
            string wantPlain = root.GetProperty("object").GetProperty("content_id_hex").GetString()!;

            // the audience-bearing content id (body WITH field 13) reproduces the oracle
            Assert.Equal(wantAudience, Hex(AudienceObject().ContentId()));
            // omit-when-empty is additive: a no-audience object reproduces the BASE draft-00 content id
            Assert.Equal(wantPlain, Hex(PlainObject().ContentId()));
            // and the two must actually differ (the audience field changes the identity)
            Assert.NotEqual(Hex(AudienceObject().ContentId()), Hex(PlainObject().ContentId()));
        }

        [Fact]
        public void SignVerifyRoundtripPreservesAudience()
        {
            byte[] seed = new byte[32];
            for (int i = 0; i < seed.Length; i++) { seed[i] = 0x2a; }
            byte[] pk = Cose.MldsaKeygen("ML-DSA-65", seed);
            string signerId = Identity.SignerId(Alg, pk);
            var obj = new Envelope.Object(
                kind: 1, channel: 4, signer: Encoding.UTF8.GetBytes(signerId), created: 1785000000000L,
                effect: 2, body: new Cbor.M(new List<Cbor.Pair> { new Cbor.Pair(new Cbor.U(1), new Cbor.T("hi")) }),
                profile: Cose.PROFILE_PUBLIC, audience: "authority-A");
            byte[] signed = Envelope.Sign(obj, Alg, seed);
            Envelope.Object got = Envelope.Verify(Cose.PROFILE_PUBLIC, Alg, pk, (c, k) => c == 4 && k == 1, signed);
            Assert.Equal("authority-A", got.Audience);
        }

        [Theory]
        [InlineData("authority-A", true)]   // consume-once, correct
        [InlineData("", false)]             // unrestricted, absent
        [InlineData("authority-A", false)]  // unrestricted, correct
        public void CheckAudiencePasses(string aud, bool consumeOnce)
        {
            Envelope.CheckAudience(GateObject(aud), "authority-A", consumeOnce); // no throw
        }

        [Theory]
        [InlineData("authority-B", true)]   // consume-once, foreign
        [InlineData("", true)]              // consume-once, absent
        [InlineData("authority-B", false)]  // foreign even when not consume-once
        public void CheckAudienceRejects(string aud, bool consumeOnce)
        {
            var ex = Assert.Throws<NaalpException>(() => Envelope.CheckAudience(GateObject(aud), "authority-A", consumeOnce));
            Assert.Equal("WrongAudience", ex.Kind);
        }

        private static byte[] Aid()
        {
            byte[] a = new byte[50];
            for (int i = 0; i < 50; i++) { a[i] = (byte)i; }
            return a;
        }

        private static void WithLedger(string authority, Action<Approval.Ledger> body)
        {
            string path = Path.GetTempFileName();
            try
            {
                using Approval.Ledger led = Approval.OpenLedger(path, authority);
                body(led);
            }
            finally
            {
                try { File.Delete(path); } catch { /* best-effort */ }
            }
        }

        [Fact]
        public void ConsumeObjectWrongAudienceRejectedNoAppend()
        {
            WithLedger("authority-A", led =>
            {
                var ex = Assert.Throws<NaalpException>(() => led.ConsumeObject(GateObject("authority-B"), Aid(), "consumer"));
                Assert.Equal("WrongAudience", ex.Kind);
                Assert.Equal(0, led.Len());
            });
        }

        [Fact]
        public void ConsumeObjectAbsentAudienceRejectedNoAppend()
        {
            WithLedger("authority-A", led =>
            {
                var ex = Assert.Throws<NaalpException>(() => led.ConsumeObject(GateObject(""), Aid(), "consumer"));
                Assert.Equal("WrongAudience", ex.Kind);
                Assert.Equal(0, led.Len());
            });
        }

        [Fact]
        public void ConsumeObjectCorrectAudienceConsumesOnce()
        {
            WithLedger("authority-A", led =>
            {
                led.ConsumeObject(GateObject("authority-A"), Aid(), "consumer");
                Assert.Equal(1, led.Len());
                var ex = Assert.Throws<NaalpException>(() => led.ConsumeObject(GateObject("authority-A"), Aid(), "consumer"));
                Assert.Equal("AlreadyConsumed", ex.Kind);
            });
        }

        [Fact]
        public void ConsumeObjectUnnamedLedgerRefuses()
        {
            WithLedger("", led =>
            {
                var ex = Assert.Throws<NaalpException>(() => led.ConsumeObject(GateObject("authority-A"), Aid(), "consumer"));
                Assert.Equal("LedgerUnsigned", ex.Kind);
            });
        }
    }
}
